pandas/core/algorithms.py

```diff
@@ -1876,21 +1876,21 @@ def _sort_tuples(values: np.ndarray) -> np.ndarray:
     return values[indexer]
 
 
-def union_with_duplicates(lvals: np.ndarray, rvals: np.ndarray) -> np.ndarray:
+def union_with_duplicates(lvals: ArrayLike, rvals: ArrayLike) -> ArrayLike:
     """
     Extracts the union from lvals and rvals with respect to duplicates and nans in
     both arrays.
 
     Parameters
     ----------
-    lvals: np.ndarray
+    lvals: ArrayLike
         left values which is ordered in front.
-    rvals: np.ndarray
+    rvals: ArrayLike
```

Comment:
if we're getting EA here, then the np.append call below is likely going to do something unwanted like cast to object

Reply:
pd_array casts back in this case

Reply:
right, but object casting is costly

Reply:
Ah got you now.